backend/src/controllers/courseController.mjs

```diff
@@ -61,20 +63,65 @@ export const getCourse = async (req, res) => {
 // Fetch all courses
 export const getAllCourses = async (req, res) => {
     try {
-        const coursesSnapshot = await db.collection(COURSES).get();  
-
-        const courses = [];
-        coursesSnapshot.forEach(doc => {
-            courses.push({ id: doc.id, ...CourseModel.fromFirestore(doc.data()) });
-        });
-
+        const coursesRef = db.collection(COURSES);
+        const coursesSnapshot = await coursesRef.get();
+
+        const courses = await Promise.all(coursesSnapshot.docs.map(async (doc) => {
+            const courseModel = CourseModel.fromFirestore(doc.data());
+            const lessons = await getCourseLessons(doc.id, coursesRef);
+            const userCourses = await getUserCourses(courseModel.courseId);
+            return {
+                id: doc.id,
+                lessons: lessons,
+                user_courses: userCourses,
+                ...courseModel
+            };
+        }));
         res.status(200).send(courses);
     } catch (error) {
         console.error('Error fetching courses:', error);
         res.status(500).send({ error: error.message });
     }
 };
 
+const getCourseLessons = async (courseId, coursesRef) => {
+    try {
+        const lessonsSnapshot = await coursesRef.doc(courseId).collection(LESSONS).get();
+
+        const lessons = [];
+        lessonsSnapshot.forEach(doc => {
+            lessons.push({ id: doc.id, ...LessonModel.fromFirestore(doc.data()) });
+        });
+
+        return lessons;
+    } catch (error) {
+        console.error('Error fetching lessons:', error);
+        return [];
+    }
+};
+
+const getUserCourses = async (courseId) => {
```

Comment:
Would recommend leaving comments above big functions like this, explaining their use. For example:
```
/* 
Given a specific course, gets all user-course objects from firestore. 
Useful for example, if you wanted to see all the students taking a given course. 
Note: Will also pull the teacher's id.
*/
 `